pkg/kubelet/kubelet.go

```diff
@@ -349,12 +369,37 @@ func (self *podWorkers) Run(podFullName string, action func()) {
 	}()
 }
 
-func makeEnvironmentVariables(container *api.Container) []string {
+func (kl *Kubelet) makeEnvironmentVariables(container *api.Container) ([]string, error) {
 	var result []string
+	// Add service environment variables.  Note:  These are added to the docker.Config, but are
+	// not included in the checksum computed by dockertools.BuildDockerName(...).  That way,
+	// we can still determine whether an api.Container is already running by its hash.
+	// (We don't want to restart a container just because some service changed.)
+	//
+	// TODO: limit env vars to same namespace services, plus kubernetes{-ro}. (Issue #3293)
+	//
+	// Note that there is a race between Kubelet seeing the pod and kubelet seeing the service.
+	// To avoid this users can: (1) wait between starting a service and starting; or (2) detect
+	// missing service env var and exit and be restarted; or (3) use DNS instead of env vars
+	// and keep trying to resolve the DNS name of the service (recommended).
+	serviceEnv, err := kl.getServiceEnvVarMap()
+	if err != nil {
+		return result, err
+	}
 	for _, value := range container.Env {
+		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
+		// So, the master may set service env vars, or kubelet may.  In case both are doing
+		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
+		// env vars.
+		// TODO: remove this net line once all platforms use apiserver+Pods.
+		delete(serviceEnv, value.Name)
```

Comment:
and of course item 3 should read:
"pods and services from APISERVER..."

On Mon, Jan 12, 2015 at 9:16 AM, Eric Tune <etune@google.com> wrote:

> You are right, my mistake.  There is no use case for service from etcd.
> The 3 use cases I was thinking of are :
>   1. boundPod from etcd with service env var set in apiserver (for
> lagging cloud providers)
>   2. boundPod from etcd with service env var set in apiserver, plus
> kubelet watching services via api. (on leading cloud provider, after your
> PR goes in but mine not in yet)
>   3. pod and services from etcd (leading cloud providers, after mine goes
> in, after yours).
>
> What you described should work great for all those cases.
>
> On Mon, Jan 12, 2015 at 8:41 AM, Paul Morie <notifications@github.com>
> wrote:
>
>> In pkg/kubelet/kubelet.go
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376#discussion-diff-22802244>
>> :
>>
>> >  	for _, value := range container.Env {
>> > +		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
>> > +		// So, the master may set service env vars, or kubelet may.  In case both are doing
>> > +		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
>> > +		// env vars.
>> > +		// TODO: remove this net line once all platforms use apiserver+Pods.
>> > +		delete(serviceEnv, value.Name)
>>
>> @erictune <https://github.com/erictune> Got it, that makes sense. I will
>> make my PR implement the following:
>>
>>    1. Keep the changes to set the service vars from the correct
>>    namespace in BoundPodFactory
>>    2. Set the service variables in the kubelet, deduplicating variables
>>    set in the API server if necessary.
>>
>> One remaining question about this sentence:
>>
>> Basically three modes have to work right: both boundPod and Service from
>> etcd;
>> boundPod etcd but service api; or pods (not bound) and service from api.
>>
>> It seems to me that this refers to the mechanism by which the kubelet
>> reads the service information. Is it accurate that we have to support both
>> direct etcd and via etcd? I didn't realize there was a use-case to read
>> directly from etcd.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3376/files#r22802244>
>> .
>>
>
>